File: MB-Secretaria/providers_py/provider_audit.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone

from .types import ProviderRequest, ProviderResponse


@dataclass(frozen=True)
class ProviderAuditEvent:
    trace_id: str
    provider: str
    model: str
    request_type: str
    modality: str
    latency_ms: int
    timeout: bool
    fallback_used: bool
    retry_count: int
    phi_detected: bool
    blocked_by_policy: bool
    created_at: str


def build_provider_audit_event(
    request: ProviderRequest,
    response: ProviderResponse,
    *,
    timeout: bool = False,
    phi_detected: bool = False,
    blocked_by_policy: bool = False,
) -> ProviderAuditEvent:
    return ProviderAuditEvent(
        trace_id=request.trace_id,
        provider=response.provider_name,
        model=response.model_name,
        request_type=request.request_type,
        modality=request.modality,
        latency_ms=response.latency_ms,
        timeout=timeout,
        fallback_used=response.fallback_used,
        retry_count=response.retry_count,
        phi_detected=phi_detected,
        blocked_by_policy=blocked_by_policy,
        created_at=datetime.now(timezone.utc).isoformat(),
    )

File: MB-Chat/providers/provider-audit.ts
import type { ProviderAuditEvent, ProviderRequest, ProviderResponse } from "./types";

export function buildProviderAuditEvent(input: {
  request: ProviderRequest;
  response: ProviderResponse;
  timeout?: boolean;
  phi_detected?: boolean;
  blocked_by_policy?: boolean;
}): ProviderAuditEvent {
  return {
    trace_id: input.request.trace_id,
    provider: input.response.provider_name,
    model: input.response.model_name,
    request_type: input.request.request_type,
    modality: input.request.modality,
    latency_ms: input.response.latency_ms,
    timeout: Boolean(input.timeout),
    fallback_used: input.response.fallback_used,
    retry_count: input.response.retry_count,
    phi_detected: Boolean(input.phi_detected),
    blocked_by_policy: Boolean(input.blocked_by_policy),
    created_at: new Date().toISOString(),
  };
}

export type ProviderAuditSink = {
  record(event: ProviderAuditEvent): void | Promise<void>;
};
